### PR TITLE
Adds self sigining cert to dex-ingress

### PR DESCRIPTION
**Pull request by krishnadurai**
**Which issue is resolved by this Pull Request:**
Addresses partially: kubeflow/kfctl#6

**Description of your changes:**
This PR contains a K8s job to generate self-signed certificates for Dex based ingress endpoint, which due to OIDC Authservice, is Istio Ingressgateway Service. This uses cert-manager.

Run `kustomize build dex-auth/dex-ingress/base | kubectl apply -f -` to test.

**Params**
domain: domain for ingress-gateway

**Note**

Certificate for LetsEncrypt not implemented since there are multiple available types for [Challenge Types](https://letsencrypt.org/docs/challenge-types/). We can expect the users of this manifest to overlay for letsencrypt based on their deployment.

**Checklist:**
- [x] Unit tests have been rebuilt: 
    1. `cd manifests/tests`
    2. `make generate`
    3. `make test`

/assign @krishnadurai 
/cc @jlewi @yanniszark

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/manifests/362)

### CODE REVIEW

**Comment by yanniszark** [comment]
@krishnadurai thanks a lot for this PR.

Unfortunately, cert-manager v0.4.0, which we currently use, does not have good support for self-signed certificates (doesn't support the `isCA` field).

This is an opportunity to also update to a later version of cert-manager (v0.10.0).
This will also let us automate certificate creation for webhooks (like the PodDefault webhook in `manifests/admission-webhook` and also for our various controller webhooks like the conversion webhooks @jlewi).

In addition, we should follow a more organized structure like:
```bash
├── base           # controllers only
├── overlays           # controllers only
│   ├── lets-encrypt
│   ├── self-signed
```bash

so that users can easily select what they want.

@krishnadurai @jlewi what do you think?

**Comment by krishnadurai** [comment]
@yanniszark good catch. Looks like it has been added for 0.5.0.
https://github.com/jetstack/cert-manager/commit/8d3d095a294ae47b5f79fdb274a1638f48ef972c

I'll just add another PR updating cert-manager to 0.10.0 which will unblock this PR.

I think adding overlays should be fine as users can choose which issuer they want. The only problem I forsee is in the KfDef application definition. For each application, a user might have to configure their 'Certificate' resource's issuer to either of self-signed or letsencrypt.
Could we introduce a global declaration for a default certificate issuer in KfDef?

**Comment by yanniszark** [comment]
> Should we introduce a global declaration for a default certificate issuer in KfDef?

I don't think that's necessary at first.
The individual applications can expose the issuer name as a parameter and the KfDef will specify the right one for the use case.

**Comment by krishnadurai** [comment]
I'll make these changes soon.

**Comment by krishnadurai** [comment]
@yanniszark PTAL at #369. It covers the upgrade. I should be able to rebase this branch on #369 once it is merged in master to address your changes here.

File: dex-auth/dex-ingress/base/certificate.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: istio-ingress-crt
+spec:
+  secretName: istio-ingressgateway-certs
+  domains:
+  - $(domain)
+  commonName: "kubeflow-root-ca"
```
[the remark this answers is not in the document]
Comment (krishnadurai) [reply dex-auth/dex-ingress/base/certificate.yaml]: We could change this to "istio-ingressgateway-root-ca" -  would make more sense.

File: dex-auth/dex-ingress/base/certificate.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+apiVersion: cert-manager.io/v1alpha2
+kind: Certificate
+metadata:
+  name: istio-ingress-crt
+spec:
+  secretName: istio-ingressgateway-certs
```
[the remark this answers is not in the document]
Comment (krishnadurai) [reply dex-auth/dex-ingress/base/certificate.yaml]: Nice suggestion @yanniszark, I was thinking on the lines that these resources belong to the Dex setup. We could keep this independent from Dex as an overlay to `istio/istio` folder calling this as ingressgateway-self-signed-cert.

**Comment by krishnadurai** [comment]
@yanniszark I've addressed your review comments.

**Comment by johnugeorge** [comment]
\lgtm

**Comment by jlewi** [comment]
/lgtm
/approve
/hold
in case you want to wait for @yanniszark

**Comment by krishnadurai** [comment]
This overlay has namespace set as `istio-system` and the base resources need to be in `kubeflow`.

**Comment by krishnadurai** [comment]
Moved `ingressgateway-self-signed-cert` to its own base folder since:
1. It would have to be deployed in `istio-system` namespace, rather than `kubeflow` namespace for istio/istio components on which this was earlier overlayed.
1. Moving this as an overlay on top of istio/istio-install didn't seem appropriate with multiple istio versions being in picture.

**Pull request merged**